src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,656 @@
+            var methodSymbolInfo = GetMethodSymbolFromExpression(document, expression, cancellationToken)!;
+            var methodCallSites = await FindCallSitesAsync(document, methodSymbolInfo, cancellationToken).ConfigureAwait(false);
+
+            var updatedCallSitesSolution = await RewriteCallSitesAsync(document, expression, methodCallSites, allOccurrences, parameterName, cancellationToken).ConfigureAwait(false);
+            return updatedCallSitesSolution;
+        }
+
+        /// <summary>
+        /// Introduces a new parameter and a new method that calls upon the updated method so that all
+        /// refactorings are coalesced to one location
+        /// </summary>
+        public async Task<Solution> IntroduceParameterForTrampolineAsync(SemanticDocument document, TExpressionSyntax expression, bool allOccurrences, bool trampoline, CancellationToken cancellationToken)
+        {
+            var parameterName = GetNewParameterName(document, expression, cancellationToken);
+            var methodSymbolInfo = GetMethodSymbolFromExpression(document, expression, cancellationToken)!;
+            var methodCallSites = await FindCallSitesAsync(document, methodSymbolInfo, cancellationToken).ConfigureAwait(false);
+            var updatedInvocationsSolution = await RewriteCallSitesWithNewMethodAsync(document,
+                expression, methodSymbolInfo, allOccurrences, parameterName, methodCallSites, trampoline, cancellationToken).ConfigureAwait(false);
+            return updatedInvocationsSolution;
+        }
+
+        protected static string GetNewParameterName(SemanticDocument document, TExpressionSyntax expression, CancellationToken cancellationToken)
+        {
+            if (IsExpressionRemovable(document, expression))
+            {
+                var syntaxFacts = document.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+                return syntaxFacts.GetIdentifierOfVariableDeclarator(expression.Parent!.Parent!).ValueText;
+            }
+
+            var semanticFacts = document.Document.GetRequiredLanguageService<ISemanticFactsService>();
+            return semanticFacts.GenerateNameForExpression(document.SemanticModel, expression, capitalize: false, cancellationToken);
+        }
+
+        protected static bool IsExpressionRemovable(SemanticDocument document, TExpressionSyntax expression)
+        {
+            var syntaxFacts = document.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+            return expression.Parent is not null && expression.Parent.Parent is not null && syntaxFacts.IsVariableDeclarator(expression.Parent.Parent);
+        }
+
+        /// <summary>
+        /// Gets the method symbol the expression is enclosed within
+        /// </summary>
+        protected IMethodSymbol? GetMethodSymbolFromExpression(SemanticDocument annotatedSemanticDocument, TExpressionSyntax annotatedExpression, CancellationToken cancellationToken)
```

Comment:
can we not just use SemanticModel.GetEnclsoingSymbol, and just walk up that till we find an IMethodSYmbol?